### PR TITLE
e2e tests for storage policy support in Kubernetes

### PR DESCRIPTION
**Pull request by BaluDontu**
This PR covers e2e test cases for vSphere storage policy support in Kubernetes - #46176.

The following test scenario have been implemented.
- Specify only SPBM storage policy name.
     - Verify if the disk is provisioned on a compatible datastore with max free space.
- Specify a storage policy name which is not defined on VC.
    - Verify if PVC create errors out that no pbm profile with this policy is found.
- Specify both SPBM storage policy name and VSAN capabilities together.
    - Verify if PVC create errors out that you can't use both SPBM policy name with VSAN capabilities. You can only specify one.
- Specify SPBM storage policy name with user specified datastore which is non-compatible.
   - Verify if PVC create errors out that it can't provision a disk on a non-compatible datastore.

@jeffvance @divyenpatel

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by BaluDontu** [comment]
@saad-ali @jeffvance: Can you start the tests on this PR.

**Comment by jeffvance** [comment]
@k8s-bot test this

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 1]
```diff
@@ -72,7 +85,7 @@ const (
    8. Delete PVC, PV and Storage Class
 */
 
-var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Volume]", func() {
+var _ = framework.KubeDescribe("Storage policy support for dynamic provisioning [Volume]", func() {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: If this is a vSphere specific test suite then "vsphere" should appear in the description text.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 2]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
 		errorMsg := "The specified datastore: \\\"" + VmfsDatastore + "\\\" is not a VSAN datastore. " +
 			"The policy parameters will work only with VSAN Datastore."
-		framework.Logf("errorMsg: %+q", errorMsg)
+		if !strings.Contains(err.Error(), errorMsg) {
+			Expect(err).NotTo(HaveOccurred())
+		}
+	})
+
+	It("verify an existing and compatible SPBM policy is honored for dynamically provisioned pvc using storageclass", func() {
+		By(fmt.Sprintf("Invoking Test for SPBM policy: %s", os.Getenv("VSPHERE_SPBM_GOLD_POLICY")))
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: What if these env vars are not set?

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 1 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: Sure. I will make the change.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 3]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
+		scParameters[SpbmStoragePolicy] = os.Getenv("VSPHERE_SPBM_GOLD_POLICY")
+		scParameters[DiskFormat] = "thin"
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: What about the `ThinDisk` const?

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 4]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
+	})
+
+	It("verify if a SPBM policy is not honored on a non-compatible datastore for dynamically provisioned pvc using storageclass", func() {
+		By(fmt.Sprintf("Invoking Test for SPBM policy: %s and datastore: %s", os.Getenv("VSPHERE_SPBM_TAG_POLICY"), VsanDatastore))
+		scParameters[SpbmStoragePolicy] = os.Getenv("VSPHERE_SPBM_TAG_POLICY")
+		scParameters[Datastore] = VsanDatastore
+		scParameters[DiskFormat] = ThinDisk
+		framework.Logf("Invoking Test for SPBM storage policy on a non-compatible datastore: %+v", scParameters)
+		err := invokeInvalidPolicyTestNeg(client, namespace, scParameters)
+		// This will make sure err is always returned.
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: comment doesn't make sense to me.  The `Expect()` below causes the test to fail if err is nil. The failure will show that the test expected err to be nil and it wasn't. If err != nil nothing happens. In either case err is not returned.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 5]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
+		Expect(err).To(HaveOccurred())
+		errorMsg := "User specified datastore: \\\"" + VsanDatastore + "\\\" is not compatible with the storagePolicy: \\\"" + os.Getenv("VSPHERE_SPBM_TAG_POLICY") + "\\\""
+		if !strings.Contains(err.Error(), errorMsg) {
+			Expect(err).NotTo(HaveOccurred())
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: An error *did* occur but not the error you were expecting. In this case you should add the extra string parm, shown above, to include what type of error you were expecting.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 6]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
+	})
+
+	It("verify if a non-existing SPBM policy is not honored for dynamically provisioned pvc using storageclass", func() {
+		By(fmt.Sprintf("Invoking Test for SPBM policy: %s", BronzeStoragePolicy))
+		scParameters[SpbmStoragePolicy] = BronzeStoragePolicy
+		scParameters[DiskFormat] = ThinDisk
+		framework.Logf("Invoking Test for non-existing SPBM storage policy: %+v", scParameters)
+		err := invokeInvalidPolicyTestNeg(client, namespace, scParameters)
+		// This will make sure err is always returned.
+		Expect(err).To(HaveOccurred())
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: same Expect comments as above...

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 2 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: I will fail the test case, if the required envt. variable is empty. Something like this.

Expect(os.Getenv("VSPHERE_SPBM_GOLD_POLICY")).NotTo(BeEmpty()).

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 3 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: I will make the change.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 7]
```diff
@@ -181,21 +194,72 @@ var _ = framework.KubeDescribe("VSAN policy support for dynamic provisioning [Vo
...
+
+	It("verify an if a SPBM policy and VSAN capabilities cannot be honored for dynamically provisioned pvc using storageclass", func() {
+		By(fmt.Sprintf("Invoking Test for SPBM policy: %s with VSAN storage capabilities", os.Getenv("VSPHERE_SPBM_GOLD_POLICY")))
+		scParameters[SpbmStoragePolicy] = os.Getenv("VSPHERE_SPBM_GOLD_POLICY")
+		scParameters[Policy_DiskStripes] = DiskStripesCapabilityVal
+		scParameters[DiskFormat] = ThinDisk
+		framework.Logf("Invoking Test for SPBM storage policy and VSAN capabilities together: %+v", scParameters)
+		err := invokeInvalidPolicyTestNeg(client, namespace, scParameters)
+		// This will make sure err is always returned.
+		Expect(err).To(HaveOccurred())
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: ditto...

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 2 above)
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: OK. I assume there are changes you need to cause these vars to be set in the CI test environment?

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 8]
```diff
@@ -204,7 +268,7 @@ func invokeValidVSANPolicyTest(f *framework.Framework, client clientset.Interfac
 	Expect(err).NotTo(HaveOccurred())
 	defer framework.DeletePersistentVolumeClaim(client, pvclaim.Name, namespace)
 
-	pvclaims := make([]*v1.PersistentVolumeClaim, 1)
+	pvclaims := make([]*v1.PersistentVolumeClaim, 0)
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: why this change?

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 2 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: True.

**Comment by jeffvance** [comment]
/approve
@k8s-bot ok to test

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 8 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: I only need pvclaims of size 1.  The following code pvclaims := make([]*v1.PersistentVolumeClaim, 0) does that. 
Having make([]*v1.PersistentVolumeClaim, 1), creates a pvclaims of size 2 which is not needed.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 9]
```diff
@@ -49,10 +54,16 @@ const (
 	StripeWidthCapabilityVal                   = "2"
 	DiskStripesCapabilityInvalidVal            = "14"
 	HostFailuresToTolerateCapabilityInvalidVal = "4"
+	VsphereSpbmGoldPolicy                      = "gold"
```
Comment (divyenpatel) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: VsphereSpbmGoldPolicy and VsphereSpbmTagbasedPolicy is not used in the test.

File: test/e2e/storage/vsphere_volume_vsan_policy.go [hunk 10]
```diff
@@ -41,6 +42,10 @@ const (
 	Policy_CacheReservation                    = "cacheReservation"
 	Policy_ObjectSpaceReservation              = "objectSpaceReservation"
 	Policy_IopsLimit                           = "iopsLimit"
+	DiskFormat                                 = "diskformat"
+	ThinDisk                                   = "thin"
+	SpbmStoragePolicy                          = "storagepolicyname"
```
Comment (divyenpatel) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: SpbmStoragePolicy not used in the test.

**Review by divyenpatel — CHANGES_REQUESTED** [review]

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 9 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: Will remove this.

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 10 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: It is being used in the last 4 test cases for SPBM policy mgmt.

**Comment by BaluDontu** [comment]
@jeffvance : I have addresses all your comments, can you check it out and provide an /lgtm if everything looks good.

**Comment by jeffvance** [comment]
@k8s-bot ok to test
@BaluDontu it's a good practice to make review changes in a new commit so that the reviewer can see just the changes you made based on his/her comments. Then when all is good to squash to a single commit.  Just fyi...

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 8 above)
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: 0 is the default initial size (and capacity). So it is more idiomatic to omit the size in this case.
However, since you `append` to the slice I don't see how setting its initial size to 1 cause the slice to contain 2 elements. When the initial size is smaller than the size needed to do an append then the slice is copied to a new location, incremented in size, and the append occurs. Defining its size as 0 skip this step in the append operation.

**Comment by BaluDontu** [comment]
@jeffvance : Since its a small change, I have squashed the commits together. I will make sure I will add it as a separate commit next time.

File: test/e2e/storage/vsphere_volume_vsan_policy.go (hunk 8 above)
Comment (BaluDontu) [inline test/e2e/storage/vsphere_volume_vsan_policy.go]: I have addressed this in the latest commit.

**Comment by jeffvance** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by jeffvance** [comment]
@k8s-bot ok to test
/lgtm

**Comment by BaluDontu** [comment]
@saad-ali  @jingxu97 : Can anyone of you provide an approval for this PR.

**Comment by jeffvance** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by jeffvance** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
